errors.go

```diff
@@ -92,35 +93,37 @@ func (w WriteErrors) Error() string {
 //   - If a non-duplicate-job error is encountered it is returned.
 //   - If every error is a duplicate job error and at least one of them is a duplicate scope error a duplicate scope error is returned.
 //   - If every error is a duplicate job error and none of them is a duplicate scope error a duplicate job error is returned.
-func (w WriteErrors) Cause() error {
-	if len(w.OtherErrors) > 0 {
+func (w *writeErrors) Cause() error {
+	if len(w.otherErrors) > 0 {
 		return w
 	}
-
-	var duplicateError error
-	if len(w.DuplicateScopeErrors) > 0 {
-		duplicateError = MakeDuplicateJobScopeError(w)
+	if len(w.duplicateScopeErrors) > 0 {
+		return MakeDuplicateJobScopeError(w)
 	}
-	if len(w.DuplicateJobErrors) > 0 {
-		duplicateError = MakeDuplicateJobError(w)
+	if len(w.duplicateJobErrors) > 0 {
+		return MakeDuplicateJobError(w)
 	}
 
-	return duplicateError
+	return nil
 }
 
-// CollateWriteErrors collates errors into a [WriteErrors].
-func CollateWriteErrors(errs []error) WriteErrors {
-	var writeErrs WriteErrors
+// CollateWriteErrors collates errors into a [writeErrors].
+func CollateWriteErrors(errs []error) *writeErrors {
```

Comment:
Generally, an exported function returning an unexported type is undesirable because the caller can't use it in any way other than for its exported methods. Can we just make this return an `error`?

Reply:
done.
Though I'm not sure I follow: if it's an `error` it's even more limiting

Reply:
Some Go linters apparently flag it as difficult to use/discouraged. There's been [some discussions](https://www.reddit.com/r/golang/comments/xurmnl/return_unexported_type/) about this before, and the general gist seems to be "if an exported method needs to return something a user wants to use, exporting just the type name is a good workaround". The Go standard library does this a bunch when they want to expose a type you can look at but can't modify the internals.